### ISSUE TITLE
Is it possible to implement Operator Fusion in ggml?

### ISSUE THREAD

**Issue by bobqianic** [ggerganov/llama.cpp#3124]
A month back, I came across this fascinating paper titled "[DNNFusion: Accelerating Deep Neural Networks Execution with Advanced Operator Fusion](https://arxiv.org/pdf/2108.13342.pdf)." In it, the authors claim to have achieved several-fold speedup and a significant reduction in inference latency with their implementation. They've provided a very detailed methodology, making it seem worthwhile to give it a shot.

I opened an issue on the ggml repository a month ago, but unfortunately, I haven't received any responses yet :(
You can find it here: [ggml issue #451](https://github.com/ggerganov/ggml/issues/451).

**Comment by goerch** [comment]
@ikawrakow is [playing](https://github.com/ggerganov/llama.cpp/pull/3121) with the idea, it seems.

**Comment by bobqianic** [comment]
#3121

**Comment by bobqianic** [comment]
> @ikawrakow is [playing](https://github.com/ggerganov/llama.cpp/pull/3121) with the idea, it seems.

Thanks!

**Comment by ikawrakow** [comment]
For the LLMs supported by `llama.cpp` one cannot get anywhere near the speedups claimed in this paper. Example: token generation (TG) for 7B on Metal takes about 15.9 ms/token on my M2 Max. If I turn matrix multiplications into no-ops, it becomes 3.4 ms/token. I.e., if I somehow magically managed to fuse all operations that are not matrix multiplications in such a way that they take zero time, TG time will become 15.9 - 3.4 = 12.5 ms, so I would gain a 15.9/12.5 = 1.27 fold speedup, so nowhere near the speedups claimed in the paper. In practice, we are probably looking at not more than a few percent gain in performance if all operations that can be fused were fused. The more important gain would be the reduction in required memory for storing intermediate results, which is even more important for training than it is for inference (but training has not been a focus of this project so far). But not all fused operations lead to memory reduction. For instance, the POC in PR #3121 leads to zero memory reduction for inference as the 3 operations being fused are all done in-place (but it would remove 2 gradient tensors in the case of training, at the added cost of now having to implement a more complicated derivative).
